### PR TITLE
Add logits processor to spec gen

### LINKED ISSUE #1255: [mlx_lm] Use logits_processors in speculative_generate_step
Right now, `logits_processors` is unused in `speculative_generate_step` https://github.com/ml-explore/mlx-examples/blob/9c2ef38d4d7a6c6afc1f9c90cb26cadb95f7343e/llms/mlx_lm/utils.py#L335 so things like repetition penalty and whatever else one may want to do in a logits processor is not being applied when using speculative decoding.

### PR DESCRIPTION
**Pull request by awni**
Closes #1255

### CODE REVIEW

File: llms/mlx_lm/utils.py [hunk 1]
```diff
@@ -390,17 +390,41 @@ def speculative_generate_step(
...
+                    y = y[: -(n_predict - 1)]
+                for i in range(n_predict):
+                    prev_tokens = (
+                        mx.concat([prev_tokens, y]) if prev_tokens is not None else y
+                    )
+                    y, logprobs = _process_and_sample(
+                        prev_tokens, logits[:, i : i + 1, :]
+                    )
+                    out_y.append(y)
+                    out_logprobs.append(logprobs)
```
Comment (awni) [inline llms/mlx_lm/utils.py:422]: The unrolling here is a bit unfortunate.. but the logits processors can be autoregressive themselves (e.g. the `repetition_penalty`.

**Review by angeloskath — APPROVED** [review]
👍

Honestly I don't mind the unrolling, it is quite natural. The only annoying thing is the manhandling of the `prev_tokens` but I don't see anything that is clearly better.
